core/google/cloud/exceptions.py

```diff
@@ -55,5 +55,6 @@
 BadGateway = exceptions.BadGateway
 ServiceUnavailable = exceptions.ServiceUnavailable
 GatewayTimeout = exceptions.GatewayTimeout
+RetryError = exceptions.RetryError
```

Comment:
Please do not add new aliases to this file (I am eventually going to deprecate this)

Reply:
Thanks. Almost there.  Just not sure how to do the checking for ServiceUnavailable (the exception that we want).  Here is the retry logic with the predicate you are recommending:

```
retry_commit = functools.partial(
    _call_mutate_row, 
     table=self._table, 
     request_pb=request_pb)

retry_ = retry.Retry(
    predicate=retry.if_exception_type(exceptions.from_grpc_error),
    deadline=30)

retry_(retry_commit)()
```
Where does the checking for exceptions.ServiceUnavailable go?

Reply:
@jonparrott @sduskis The requested changes were made but there were two issues: 1) call_mutate_row cannot be made private since it is then not visible in the namespace needed by the functools.partial code, 2) there is NO mapping from exceptions.GrpcRendezvous to exceptions.ServiceUnavailable, which is the original problem we identified, and therefore exceptions.from_grpc_error does not create exceptions.ServiceUnavailable, so we have had to revert to retry on Rendezvous exceptions. Please see last commit.

It does seem too complex to have to add a custom predicate and also seems like too much deep knowledge required to know we have to remap the Rendezvous exception to ServiceUnavailable ourselves.  That is why we were really hoping from the beginning that this mapping would have been done by the layers below us.  Even if this were done in those layers manually for this one requested mapping, that would be a big help.

Reply:
> 1) call_mutate_row cannot be made private since it is then not visible in the namespace needed by the functools.partial code

I don't understand, you can absolutely bind partial to a private method, e.g.:

```python
retry_commit = functools.partial(
   _call_mutate_row,
   ...)
```

> 2) there is NO mapping from exceptions.GrpcRendezvous to exceptions.ServiceUnavailable, which is the original problem we identified, and therefore exceptions.from_grpc_error does not create exceptions.ServiceUnavailable, so we have had to revert to retry on Rendezvous exceptions. Please see last commit.

Yes there is, it's right here: https://github.com/GoogleCloudPlatform/google-cloud-python/blob/7714d0cfbbf2da6aff45daef3bba8443927afd6f/api_core/google/api_core/exceptions.py#L313

> It does seem too complex to have to add a custom predicate and also seems like too much deep knowledge required to know we have to remap the Rendezvous exception to ServiceUnavailable ourselves.

It's really not, other clients do this as well.

> That is why we were really hoping from the beginning that this mapping would have been done by the layers below us. Even if this were done in those layers manually for this one requested mapping, that would be a big help.

Once we get the autogen update for bigtable in, there will be automatic mapping to exception classes and automatic retries. This PR may not even been needed in that case.

Reply:
@jonparrott We made a unit test case to step through the exception mapping code and it now seems to work.  Your help is very much appreciated.  Running our endurance tests now. If they handle the predicate, then I think the latest commit works and satisfies your review. Thanks!

Reply:
This still needs to be removed.